activemount.go

```diff
@@ -0,0 +1,152 @@
+// actually requires to be unmounted (as an overlay fs unmount). It will check the number of times the container
+// has been requested to mount the volume in the file named `requestId` and decrease the number, when the number
+// reaches zero it will delete the `requestId` file since this container no longer mounts the volume. It will
+// also check if other containers are mounting this volume by checking for other files in the active mounts folder.
+// Parameters:
+//
+//	requestName: Name of the volume to be unmounted
+//	requestID: ID of the container requesting the unmount
+//	activemountsdir: Folder where Docker-On-Top mounts are tracked.
+//
+// Return:
+//
+//	doUnmountFs: Returns true if there are not other usages of this volume and the filesystem can be unmounted.
+//	err: If the function encountered an error, the error itself, nil if everything went right.
+func (d *DockerOnTop) deactivateVolume(requestName string, requestId string, activemountsdir lockedFile) (bool, error) {
+
+	dirEntries, readDirErr := activemountsdir.ReadDir(2) // Check if there is any _other_ container using the volume
+	if errors.Is(readDirErr, io.EOF) {
+		// If directory is empty, unmount overlay and clean up
+		return true, fmt.Errorf("there are no active mount files and one was expected. Unmounting")
+	} else if readDirErr != nil {
+		return false, fmt.Errorf("failed to list activemounts/ %v", readDirErr)
+	}
+
+	otherVolumesPresent := len(dirEntries) > 1 || dirEntries[0].Name() != requestId
+	var activeMountInfo activeMount
+
+	activemountFilePath := d.activemountsdir(requestName) + requestId
+	file, err := os.Open(activemountFilePath)
+
+	if err == nil {
+		payload, readErr := io.ReadAll(file)
+		closeErr := file.Close()
+		if readErr != nil {
+			return false, fmt.Errorf("active mount file %s has been opened but cannot be read %w, the filesystem won't be unmounted", activemountFilePath, readErr)
+		} else if closeErr != nil {
+			return false, fmt.Errorf("active mount file %s has been opened but cannot be closed %w, the filesystem won't be unmounted", activemountFilePath, readErr)
+		}
+		unmarshalErr := json.Unmarshal(payload, &activeMountInfo)
+		if unmarshalErr != nil {
+			return false, fmt.Errorf("active mount file %s contents are invalid %w, the filesystem won't be unmounted", activemountFilePath, unmarshalErr)
+		}
+	} else if os.IsNotExist(err) {
+		return !otherVolumesPresent, fmt.Errorf("the active mount file %s was expected but is not there %w, the filesystem won't be unmounted", activemountFilePath, err)
+	} else {
+		return false, fmt.Errorf("the active mount file %s could not be opened %w, the filesystem won't be unmounted", activemountFilePath, err)
+	}
+
+	activeMountInfo.UsageCount--
+
+	if activeMountInfo.UsageCount == 0 {
+		err := os.Remove(activemountFilePath)
+		if err != nil {
+			return false, fmt.Errorf("the active mount file %s could not be deleted %w, the filesystem won't be unmounted", activemountFilePath, err)
```

Comment:
For consistency with an earlier error, indicating action ("unmount<i>ing</i>" / "refus<i>ing</i> to unmount")

```suggestion
			return false, fmt.Errorf("the active mount file %s could not be deleted: %w. Refusing to unmount", activemountFilePath, err)
```

Reply:
Instead of changing all messages saying "the filesystem won't be unmounted" I changed the first one that said "Unmount" to "the filesystem will be unmounted (which now is a warning).